crates/services/gas_price_service/src/v1/da_source_adapter/service.rs

```diff
@@ -69,13 +66,15 @@ where
 {
     const NAME: &'static str = "DaBlockCostsService";
 
-    type SharedData = ();
+    type SharedData = Sender<DaBlockCosts>;
 
     type Task = Self;
 
     type TaskParams = ();
 
-    fn shared_data(&self) -> Self::SharedData {}
+    fn shared_data(&self) -> Self::SharedData {
+        self.shared_state.clone()
+    }
```

Comment:
`Receiver<DaBlockCosts>` doesn't implement `Clone` :)
we can generate a new subscription with a `Sender` of tokio::sync::broadcast

Reply:
Just, semantically, a `Sender` isn't shared data. It isn't even a means of getting shared data. It's the object that you need to expose to generate the object by which you receive shared data (a `Receiver`). 

So not only are we in blatant violation of SRP with the `SharedState` (see [God Object](https://en.wikipedia.org/wiki/God_object)), but we are also requiring the consumer to understand our underlying implementation, i.e. to get this data you must know the fields and the interfaces of the fields of the `SharedData`

Instead of having a `shared_data` getter, we should just have a `get_da_block_costs_receiver()` method (this is kinda a [ISP](https://en.wikipedia.org/wiki/Interface_segregation_principle) thing as well). But we are relying on an overly-generalized `RunnableService` abstraction that hides our business domain and requires us to rely on implementation details (as seen above).

Reply:
I don't see where it is a God object.

The shared state is your object that has the `get_da_block_costs_receiver` method. 

We can rename `SharedState` into `ServiceAccessPoint`.

Instead of exposing the service and implementing the `get_da_block_costs_receiver` on the service, we have an access point to the service via the `SharedState`. No one has access to the service except the service itself. All interactions with the service are done via its access point. 

> we are also requiring the consumer to understand our underlying implementation, i.e. to get this data you must know the fields and the interfaces of the fields of the SharedData

> we should just have a get_da_block_costs_receiver() method 

I don't see why you need to know the fields of the shared state... It is just an object with the `get_da_block_costs_receiver` method.

Reply:
> I don't see where it is a God object.

Sorry. The God object point doesn't apply here. I was thinking about our txpool service `SharedState` I was looking at recently:
```rs
pub struct SharedState<
    PSProvider,
    ConsensusParamsProvider,
    GasPriceProvider,
    WasmChecker,
    MemoryPool,
> {
    pool: TxPool<PSProvider>,
    current_height: Arc<RwLock<BlockHeight>>,
    consensus_parameters_provider: Arc<ConsensusParamsProvider>,
    gas_price_provider: Arc<GasPriceProvider>,
    wasm_checker: Arc<WasmChecker>,
    memory: Arc<MemoryPool>,
    heavy_async_processor: Arc<HeavyAsyncProcessor>,
    utxo_validation: bool,
}
```

Reply:
> All interactions with the service are done via its access point.

This should be "access _point**s**_". And those access points should be defined by the domain of the service, not alllll possilbe services. When I'm saying we should have a `get_da_block_costs_receiver()` method, I'm saying that has a Single Responsiblility: providing a receiver.

Instead, we have to call `service.shared_data().get_da_block_costs_receiver()` and we have an arbitrary constraint that `SharedData` must be `Clone`, which isn't relevant here--we don't need to clone anything.

Reply:
https://github.com/FuelLabs/fuel-core/issues/2282